src/librustc_mir/interpret/machine.rs

```diff
@@ -14,6 +14,15 @@ use super::{
     InterpretCx, PlaceTy, OpTy, ImmTy, MemoryKind,
 };
 
+/// Data returned by Machine::stack_pop,
+/// to provide further control over the popping of the stack frame
+pub struct StackPopInfo {
+    /// Whether or not the machine is currently unwinding.
+    /// This affects the target block that will be jumped
+    /// to after the stack frame is popped.
+    pub unwinding: bool
```

Comment:
Why is that needed here? Seems like whether we are running a `Return` or `Resume` terminator should be enough to know this?

Reply:
This is used by the machine implementation to indicate when to stop unwinding. We tag the `catch_unwind` stack frame in Miri, so Miri needs some way of indicating to rustc that it should stop unwinding.

Reply:
The way to stop unwinding is to jump into code that doesn't unwind.

So one thing that might make sense here is to change the `stack_pop` machine hook a bit. It could take as a parameter whether we are executing a "resume" or "return"-style pop, and it takes the `StackPopCleanup`, and it returns which block to jump to (if any). That could be used both to implement `catch_panic`, and also to make sure that CTFE execution will never do "resume".

That said, I am still not sure why `catch_panic` cannot just set up the frame such that the "unwind" edge points to the same thing as the "return" edge. But that might be a bit too hacky; clearly that stack frame needs special treatment either way.